File: cli/python/rendo_cli/contracts.py
from __future__ import annotations


RUNTIME_MODES = {"source", "managed", "hybrid"}
TEMPLATE_TYPES = {"template"}
TEMPLATE_KINDS = {"starter-template", "feature-template", "capability-template", "provider-template", "surface-template"}
TEMPLATE_ROLES = {"core", "base", "derived"}
SURFACES = {"web", "miniapp", "mobile", "desktop"}
CONFLICT_STRATEGIES = {"fail", "overwrite", "skip"}
ROLLBACK_STRATEGIES = {"safe-abort", "manual"}
AUTH_TYPES = {"none", "bearer-token"}
TEMPLATE_HOST_MODELS = {"host-root", "host-attached"}
TEMPLATE_RUNTIME_CLASSES = {"standalone-runnable", "host-attached"}


def require_keys(payload: dict, keys: list[str], label: str) -> None:
    missing = [key for key in keys if key not in payload]
    if missing:
        raise RuntimeError(f"{label} is missing required keys: {', '.join(missing)}")


def validate_version_selector(payload: dict, label: str) -> dict:
    require_keys(payload, ["min", "max"], label)
    return payload


def validate_install_plan(payload: dict, label: str = "install plan") -> dict:
    require_keys(
        payload,
        [
            "addsFiles",
            "updatesFiles",
            "deletesFiles",
            "addsEnv",
            "addsRoutes",
            "addsPages",
            "addsComponents",
            "addsMigrations",
            "addsWorkerTasks",
            "addsAdminModules",
            "requiresManualSetup",
        ],
        label,
    )
    return payload


def validate_template_compatibility(payload: dict) -> dict:
    require_keys(payload, ["cli", "registryProtocol", "hosts"], "template compatibility")
    validate_version_selector(payload["cli"], "template compatibility cli")
    validate_version_selector(payload["registryProtocol"], "template compatibility registryProtocol")
    for host in payload["hosts"]:
        require_keys(host, ["templateId", "templateKind", "minVersion", "maxVersion"], "template host compatibility")
        if host["templateKind"] not in TEMPLATE_KINDS:
            raise RuntimeError(f"invalid host template kind: {host['templateKind']}")
        if host["templateId"] is not None and not isinstance(host["templateId"], str):
            raise RuntimeError("invalid host template id")
    return payload


def validate_template_architecture(payload: dict) -> dict:
    require_keys(payload, ["standard", "hostModel", "runtimeClass", "rootPaths"], "template architecture")
    if payload["hostModel"] not in TEMPLATE_HOST_MODELS:
        raise RuntimeError(f"invalid template host model: {payload['hostModel']}")
    if payload["runtimeClass"] not in TEMPLATE_RUNTIME_CLASSES:
        raise RuntimeError(f"invalid template runtime class: {payload['runtimeClass']}")
    require_keys(
        payload["rootPaths"],
        ["agentEntrypoints", "docs", "interfaces", "implementation", "tests", "scripts", "integration", "operations", "mounts"],
        "template architecture rootPaths",
    )
    return payload


def validate_asset_integration(payload: dict | None) -> dict | None:
    if payload is None:
        return None

    require_keys(payload, ["previewSummary", "supportedHostKinds", "supportedHostTemplates", "modes"], "asset integration")
    if any(kind not in TEMPLATE_KINDS for kind in payload["supportedHostKinds"]):
        raise RuntimeError("invalid supported host kind in asset integration")
    for mode in payload["modes"]:
        require_keys(
            mode,
            ["runtimeMode", "targetRoot", "conflictStrategy", "rollbackStrategy", "changes"],
            "asset integration mode",
        )
        if mode["runtimeMode"] not in RUNTIME_MODES:
            raise RuntimeError(f"invalid asset integration runtime mode: {mode['runtimeMode']}")
        if mode["conflictStrategy"] not in CONFLICT_STRATEGIES:
            raise RuntimeError(f"invalid conflict strategy: {mode['conflictStrategy']}")
        if mode["rollbackStrategy"] not in ROLLBACK_STRATEGIES:
            raise RuntimeError(f"invalid rollback strategy: {mode['rollbackStrategy']}")
        validate_install_plan(mode["changes"], "asset integration mode changes")
    return payload


def validate_template_manifest(payload: dict) -> dict:
    require_keys(
        payload,
        [
            "schemaVersion",
            "id",
            "name",
            "version",
            "type",
            "templateKind",
            "templateRole",
            "category",
            "title",
            "description",
            "uiMode",
            "domainTags",
            "scenarioTags",
            "runtimeModes",
            "defaultProviders",
            "recommendedPacks",
            "requiredEnv",
            "toolchains",
            "lineage",
            "documentation",
            "architecture",
            "surfaceCapabilities",
            "defaultSurfaces",
            "surfacePaths",
            "supports",
            "compatibility",
            "assetIntegration",
        ],
        "template manifest",
    )
    if payload["type"] not in TEMPLATE_TYPES:
        raise RuntimeError(f"invalid template type: {payload['type']}")
    if payload["templateKind"] not in TEMPLATE_KINDS:
        raise RuntimeError(f"invalid template kind: {payload['templateKind']}")
    if payload["templateRole"] not in TEMPLATE_ROLES:
        raise RuntimeError(f"invalid template role: {payload['templateRole']}")
    if any(mode not in RUNTIME_MODES for mode in payload["runtimeModes"]):
        raise RuntimeError("invalid runtime mode in template manifest")
    require_keys(payload["lineage"], ["coreTemplate", "baseTemplate", "parentTemplate"], "template lineage")
    require_keys(
        payload["documentation"],
        ["overview", "structure", "extensionPoints", "inheritanceBoundaries", "secondaryDevelopment"],
        "template documentation",
    )
    validate_template_architecture(payload["architecture"])
    if any(surface not in SURFACES for surface in payload["surfaceCapabilities"]):
        raise RuntimeError("invalid surface in template manifest")
    if any(surface not in SURFACES for surface in payload["defaultSurfaces"]):
        raise RuntimeError("invalid default surface in template manifest")
    validate_template_compatibility(payload["compatibility"])
    validate_asset_integration(payload["assetIntegration"])
    return payload


def validate_project_manifest(payload: dict) -> dict:
    require_keys(payload, ["schemaVersion", "projectName", "surfaces", "template", "installedTemplates", "installedPacks"], "project manifest")
    if "workspaceId" in payload and (not isinstance(payload["workspaceId"], str) or not payload["workspaceId"]):
        raise RuntimeError("invalid workspace id")
    require_keys(
        payload["template"],
        ["id", "templateKind", "templateRole", "version", "runtimeMode", "createdFrom", "createdAt"],
        "project template",
    )
    if "origin" in payload:
        require_keys(
            payload["origin"],
            ["createdBy", "registry", "source", "templateId", "templateKind", "templateRole", "templateVersion", "runtimeMode"],
            "project origin",
        )
        if payload["origin"]["source"] not in {"local", "remote"}:
            raise RuntimeError(f"invalid project origin source: {payload['origin']['source']}")
        if payload["origin"]["templateKind"] not in TEMPLATE_KINDS:
            raise RuntimeError(f"invalid project origin template kind: {payload['origin']['templateKind']}")
        if payload["origin"]["templateRole"] not in TEMPLATE_ROLES:
            raise RuntimeError(f"invalid project origin template role: {payload['origin']['templateRole']}")
        if payload["origin"]["runtimeMode"] not in RUNTIME_MODES:
            raise RuntimeError(f"invalid project origin runtime mode: {payload['origin']['runtimeMode']}")
    for item in payload["installedTemplates"]:
        require_keys(
            item,
            [
                "id",
                "version",
                "templateKind",
                "templateRole",
                "runtimeMode",
                "registry",
                "source",
                "bundleDigest",
                "templateDigest",
                "targetPath",
                "installedAt",
            ],
            "installed template",
        )
        if item["templateKind"] not in TEMPLATE_KINDS:
            raise RuntimeError(f"invalid installed template kind: {item['templateKind']}")
        if item["templateRole"] not in TEMPLATE_ROLES:
            raise RuntimeError(f"invalid installed template role: {item['templateRole']}")
        if item["runtimeMode"] not in RUNTIME_MODES:
            raise RuntimeError(f"invalid installed template runtime mode: {item['runtimeMode']}")
        if item["source"] not in {"local", "remote"}:
            raise RuntimeError(f"invalid installed template source: {item['source']}")
    normalized = {
        "schemaVersion": payload["schemaVersion"],
        "projectName": payload["projectName"],
    }
    if "workspaceId" in payload:
        normalized["workspaceId"] = payload["workspaceId"]
    normalized["surfaces"] = payload["surfaces"]
    normalized["template"] = payload["template"]
    if "origin" in payload:
        normalized["origin"] = payload["origin"]
    normalized["installedTemplates"] = payload["installedTemplates"]
    normalized["installedPacks"] = payload["installedPacks"]
    return normalized


def validate_pack_manifest(payload: dict) -> dict:
    require_keys(
        payload,
        [
            "name",
            "version",
            "title",
            "type",
            "runtimeMode",
            "provider",
            "description",
            "dependencies",
            "requiredEnv",
            "permissions",
            "billing",
            "install",
        ],
        "pack manifest",
    )
    if payload["runtimeMode"] not in RUNTIME_MODES:
        raise RuntimeError(f"invalid pack runtime mode: {payload['runtimeMode']}")
    validate_install_plan(payload["install"])
    return payload


def validate_digest(payload: dict, label: str = "digest") -> dict:
    require_keys(payload, ["algorithm", "value"], label)
    if payload["algorithm"] != "sha256":
        raise RuntimeError(f"invalid digest algorithm for {label}: {payload['algorithm']}")
    return payload


def validate_template_bundle(payload: dict) -> dict:
    require_keys(payload, ["schemaVersion", "bundleFormat", "templateId", "version", "templateDigest", "manifest", "files"], "template bundle")
    if payload["bundleFormat"] != "rendo-bundle.v1":
        raise RuntimeError(f"invalid template bundle format: {payload['bundleFormat']}")
    validate_digest(payload["templateDigest"], "template bundle templateDigest")
    validate_template_manifest(payload["manifest"])
    for file in payload["files"]:
        require_keys(file, ["path", "encoding", "sha256", "content"], "template bundle file")
        if file["encoding"] != "base64":
            raise RuntimeError(f"invalid template bundle file encoding: {file['encoding']}")
    return payload


def validate_registry_auth(payload: dict) -> dict:
    require_keys(payload, ["type", "header", "scheme"], "registry auth")
    if payload["type"] not in AUTH_TYPES:
        raise RuntimeError(f"invalid registry auth type: {payload['type']}")
    return payload


def validate_registry_handshake(payload: dict) -> dict:
    require_keys(
        payload,
        [
            "schemaVersion",
            "protocolVersion",
            "registryId",
            "registryTitle",
            "apiBaseUrl",
            "auth",
            "cliCompatibility",
            "bundleFormat",
            "digestAlgorithm",
        ],
        "registry handshake",
    )
    validate_registry_auth(payload["auth"])
    validate_version_selector(payload["cliCompatibility"], "registry handshake cliCompatibility")
    if "snapshot" in payload:
        require_keys(payload["snapshot"], ["url", "digest"], "registry handshake snapshot")
        validate_digest(payload["snapshot"]["digest"], "registry handshake snapshot digest")
    if payload["bundleFormat"] != "rendo-bundle.v1":
        raise RuntimeError(f"invalid registry bundle format: {payload['bundleFormat']}")
    if payload["digestAlgorithm"] != "sha256":
        raise RuntimeError(f"invalid registry digest algorithm: {payload['digestAlgorithm']}")
    return payload


def validate_registry_snapshot(payload: dict) -> dict:
    require_keys(payload, ["schemaVersion", "catalogFormat", "generatedAt", "registry", "entries"], "registry snapshot")
    if payload["catalogFormat"] != "rendo-runtime-catalog.v1":
        raise RuntimeError(f"invalid registry snapshot format: {payload['catalogFormat']}")
    require_keys(payload["registry"], ["id", "protocolVersion", "bundleFormat", "digestAlgorithm"], "registry snapshot registry")
    if payload["registry"]["bundleFormat"] != "rendo-bundle.v1":
        raise RuntimeError(f"invalid registry snapshot bundle format: {payload['registry']['bundleFormat']}")
    if payload["registry"]["digestAlgorithm"] != "sha256":
        raise RuntimeError(f"invalid registry snapshot digest algorithm: {payload['registry']['digestAlgorithm']}")
    for entry in payload["entries"]:
        require_keys(
            entry,
            [
                "id",
                "ref",
                "aliases",
                "official",
                "templateKind",
                "templateRole",
                "version",
                "runtimeModes",
                "requiredEnv",
                "toolchains",
                "lineage",
                "architecture",
                "surfaceCapabilities",
                "defaultSurfaces",
                "surfacePaths",
                "supports",
                "compatibility",
                "assetIntegration",
                "artifacts",
            ],
            "registry snapshot entry",
        )
        if entry["templateKind"] not in TEMPLATE_KINDS:
            raise RuntimeError(f"invalid registry snapshot template kind: {entry['templateKind']}")
        if entry["templateRole"] not in TEMPLATE_ROLES:
            raise RuntimeError(f"invalid registry snapshot template role: {entry['templateRole']}")
        if any(mode not in RUNTIME_MODES for mode in entry["runtimeModes"]):
            raise RuntimeError("invalid runtime mode in registry snapshot entry")
        validate_template_architecture(entry["architecture"])
        validate_template_compatibility(entry["compatibility"])
        validate_asset_integration(entry["assetIntegration"])
        require_keys(entry["artifacts"], ["manifestPath", "templatePath", "bundlePath", "bundleDigest", "templateDigest"], "registry snapshot entry artifacts")
        validate_digest(entry["artifacts"]["bundleDigest"], "registry snapshot entry bundleDigest")
        validate_digest(entry["artifacts"]["templateDigest"], "registry snapshot entry templateDigest")
    return payload


def validate_remote_search_response(payload: dict) -> dict:
    require_keys(payload, ["registry", "results"], "remote search response")
    require_keys(payload["registry"], ["id", "protocolVersion"], "remote search response registry")
    return payload


def validate_remote_inspect_response(payload: dict) -> dict:
    require_keys(payload, ["registry", "payload"], "remote inspect response")
    require_keys(payload["registry"], ["id", "protocolVersion"], "remote inspect response registry")
    if "manifest" in payload:
        validate_template_manifest(payload["manifest"])
    if "bundle" in payload:
        require_keys(payload["bundle"], ["url", "digest", "templateDigest", "bundleFormat"], "remote inspect response bundle")
        validate_digest(payload["bundle"]["digest"], "remote inspect response bundle digest")
        validate_digest(payload["bundle"]["templateDigest"], "remote inspect response bundle templateDigest")
    return payload

File: cli/node/src/contracts.ts
import { z } from "zod";

export const runtimeModeSchema = z.enum(["source", "managed", "hybrid"]);
export type RuntimeMode = z.infer<typeof runtimeModeSchema>;

export const templateTypeSchema = z.enum(["template"]);
export const templateKindSchema = z.enum([
  "starter-template",
  "feature-template",
  "capability-template",
  "provider-template",
  "surface-template",
]);
export const templateRoleSchema = z.enum(["core", "base", "derived"]);
export const templateCategorySchema = z.string().min(1);

export const uiModeSchema = z.enum(["none", "web", "miniapp", "mobile", "multi"]);
export const surfaceSchema = z.enum(["web", "miniapp", "mobile", "desktop"]);
export const domainTagSchema = z.string().min(1);
export const scenarioTagSchema = z.string().min(1);

export const supportMatrixSchema = z.object({
  web: z.boolean(),
  miniapp: z.boolean(),
  mobile: z.boolean(),
  desktop: z.boolean(),
});

export const providerMapSchema = z.object({
  agent: z.string(),
  workflow: z.string(),
  auth: z.string(),
  db: z.string(),
  cache: z.string(),
});

export const toolchainSchema = z.object({
  name: z.string().min(1),
  version: z.string().min(1),
  role: z.string().min(1),
});

export const versionSelectorSchema = z.object({
  min: z.string().min(1).nullable(),
  max: z.string().min(1).nullable(),
});
export type VersionSelector = z.infer<typeof versionSelectorSchema>;

export const lineageSchema = z.object({
  coreTemplate: z.string().nullable(),
  baseTemplate: z.string().nullable(),
});

export const documentationLinksSchema = z.object({
  overview: z.string().min(1),
  structure: z.string().min(1),
  extensionPoints: z.string().min(1),
  inheritanceBoundaries: z.string().min(1),
  secondaryDevelopment: z.string().min(1),
});

export const surfacePathsSchema = z.record(z.string(), z.array(z.string()));

export const hostCompatibilitySchema = z.object({
  templateId: z.string().min(1).nullable(),
  templateKind: templateKindSchema,
  minVersion: z.string().min(1).nullable(),
  maxVersion: z.string().min(1).nullable(),
});
export type HostCompatibility = z.infer<typeof hostCompatibilitySchema>;

export const templateCompatibilitySchema = z.object({
  cli: versionSelectorSchema,
  registryProtocol: versionSelectorSchema,
  hosts: z.array(hostCompatibilitySchema).default([]),
});
export type TemplateCompatibility = z.infer<typeof templateCompatibilitySchema>;

export const templateManifestSchema = z.object({
  schemaVersion: z.string().min(1),
  id: z.string().min(1),
  name: z.string().min(1),
  version: z.string().min(1),
  type: templateTypeSchema,
  templateKind: templateKindSchema,
  templateRole: templateRoleSchema,
  category: templateCategorySchema,
  title: z.string().min(1),
  description: z.string().min(1),
  uiMode: uiModeSchema,
  domainTags: z.array(domainTagSchema).default([]),
  scenarioTags: z.array(scenarioTagSchema).default([]),
  runtimeModes: z.array(runtimeModeSchema).min(1),
  defaultProviders: providerMapSchema,
  recommendedPacks: z.array(z.string()),
  requiredEnv: z.array(z.string()),
  toolchains: z.array(toolchainSchema).default([]),
  lineage: lineageSchema,
  documentation: documentationLinksSchema,
  surfaceCapabilities: z.array(surfaceSchema).default([]),
  defaultSurfaces: z.array(surfaceSchema).default([]),
  surfacePaths: surfacePathsSchema.default({}),
  supports: supportMatrixSchema,
  compatibility: templateCompatibilitySchema,
  assetInstall: z
    .object({
      previewSummary: z.string().min(1),
      supportedHostKinds: z.array(templateKindSchema).default([]),
      supportedHostTemplates: z.array(z.string()).default([]),
      modes: z.array(
        z.object({
          runtimeMode: runtimeModeSchema,
          targetRoot: z.string().min(1),
          conflictStrategy: z.enum(["fail", "overwrite", "skip"]),
          rollbackStrategy: z.enum(["safe-abort", "manual"]),
          install: z.object({
            addsFiles: z.array(z.string()),
            updatesFiles: z.array(z.string()),
            deletesFiles: z.array(z.string()),
            addsEnv: z.array(z.string()),
            addsRoutes: z.array(z.string()),
            addsPages: z.array(z.string()),
            addsComponents: z.array(z.string()),
            addsMigrations: z.boolean(),
            addsWorkerTasks: z.boolean(),
            addsAdminModules: z.boolean(),
            requiresManualSetup: z.boolean(),
          }),
        }),
      ),
    })
    .nullable(),
});
export type TemplateManifest = z.infer<typeof templateManifestSchema>;

export const packTypeSchema = z.enum([
  "tool-pack",
  "workflow-pack",
  "workflow-node-pack",
  "channel-pack",
  "admin-pack",
  "data-pack",
  "agent-pack",
]);

export const installPlanSchema = z.object({
  addsFiles: z.array(z.string()),
  updatesFiles: z.array(z.string()),
  deletesFiles: z.array(z.string()),
  addsEnv: z.array(z.string()),
  addsRoutes: z.array(z.string()),
  addsPages: z.array(z.string()),
  addsComponents: z.array(z.string()),
  addsMigrations: z.boolean(),
  addsWorkerTasks: z.boolean(),
  addsAdminModules: z.boolean(),
  requiresManualSetup: z.boolean(),
});
export type InstallPlan = z.infer<typeof installPlanSchema>;

export const packManifestSchema = z.object({
  name: z.string().min(1),
  version: z.string().min(1),
  title: z.string().min(1),
  type: packTypeSchema,
  runtimeMode: runtimeModeSchema,
  provider: z.enum(["rendo", "community", "third-party"]),
  description: z.string().min(1),
  dependencies: z.array(z.string()),
  requiredEnv: z.array(z.string()),
  permissions: z.array(z.string()),
  billing: z.object({
    mode: z.string().min(1),
  }),
  install: installPlanSchema,
  official: z.boolean().default(false),
  appliesTo: z.array(z.string()).default([]),
});
export type PackManifest = z.infer<typeof packManifestSchema>;

export const installedPackSchema = z.object({
  name: z.string(),
  version: z.string(),
  runtimeMode: runtimeModeSchema,
  provider: z.string(),
  installedAt: z.string(),
});

export const installedTemplateSchema = z.object({
  id: z.string(),
  version: z.string(),
  templateKind: templateKindSchema,
  templateRole: templateRoleSchema,
  runtimeMode: runtimeModeSchema,
  registry: z.string(),
  source: z.enum(["local", "remote"]),
  bundleDigest: z.string().nullable(),
  templateDigest: z.string().nullable(),
  targetPath: z.string(),
  installedAt: z.string(),
});

export const projectManifestSchema = z.object({
  schemaVersion: z.string(),
  projectName: z.string(),
  surfaces: z.array(surfaceSchema).default([]),
  template: z.object({
    id: z.string(),
    templateKind: templateKindSchema,
    templateRole: templateRoleSchema,
    version: z.string(),
    runtimeMode: runtimeModeSchema,
    createdFrom: z.string(),
    createdAt: z.string(),
  }),
  installedTemplates: z.array(installedTemplateSchema).default([]),
  installedPacks: z.array(installedPackSchema),
});
export type ProjectManifest = z.infer<typeof projectManifestSchema>;

export const templateRegistryEntrySchema = z.object({
  id: z.string(),
  ref: z.string(),
  aliases: z.array(z.string()),
  official: z.boolean(),
  templateKind: templateKindSchema,
  templateRole: templateRoleSchema,
  manifestPath: z.string(),
  templatePath: z.string(),
});
export type TemplateRegistryEntry = z.infer<typeof templateRegistryEntrySchema>;

export const packRegistryEntrySchema = z.object({
  name: z.string(),
  ref: z.string(),
  aliases: z.array(z.string()),
  official: z.boolean(),
  manifestPath: z.string(),
  filesPath: z.string(),
});
export type PackRegistryEntry = z.infer<typeof packRegistryEntrySchema>;

export const templateRegistrySchema = z.object({
  templates: z.array(templateRegistryEntrySchema),
});

export const packRegistrySchema = z.object({
  packs: z.array(packRegistryEntrySchema),
});

export const inspectPayloadSchema = z.object({
  kind: z.string().min(1),
  id: z.string(),
  title: z.string(),
  version: z.string(),
  type: z.string(),
  templateKind: templateKindSchema.optional(),
  templateRole: templateRoleSchema.optional(),
  description: z.string(),
  category: z.string().optional(),
  uiMode: z.string().optional(),
  domainTags: z.array(domainTagSchema).optional(),
  scenarioTags: z.array(scenarioTagSchema).optional(),
  toolchains: z.array(toolchainSchema).optional(),
  documentation: documentationLinksSchema.optional(),
  surfaceCapabilities: z.array(surfaceSchema).optional(),
  defaultSurfaces: z.array(surfaceSchema).optional(),
  runtimeModes: z.array(z.string()).optional(),
  runtimeMode: z.string().optional(),
  provider: z.string().optional(),
  requiredEnv: z.array(z.string()),
  dependencies: z.array(z.string()),
  official: z.boolean(),
  install: installPlanSchema.optional(),
  compatibility: templateCompatibilitySchema.optional(),
  assetInstall: templateManifestSchema.shape.assetInstall.optional(),
});
export type InspectPayload = z.infer<typeof inspectPayloadSchema>;

export const digestSchema = z.object({
  algorithm: z.literal("sha256"),
  value: z.string().min(1),
});
export type Digest = z.infer<typeof digestSchema>;

export const templateBundleFileSchema = z.object({
  path: z.string().min(1),
  encoding: z.literal("base64"),
  sha256: z.string().min(1),
  content: z.string().min(1),
});
export type TemplateBundleFile = z.infer<typeof templateBundleFileSchema>;

export const templateBundleSchema = z.object({
  schemaVersion: z.string().min(1),
  bundleFormat: z.literal("rendo-bundle.v1"),
  templateId: z.string().min(1),
  version: z.string().min(1),
  templateDigest: digestSchema,
  manifest: templateManifestSchema,
  files: z.array(templateBundleFileSchema),
});
export type TemplateBundle = z.infer<typeof templateBundleSchema>;

export const registryAuthSchema = z.object({
  type: z.enum(["none", "bearer-token"]),
  header: z.string().min(1),
  scheme: z.string().min(1).nullable(),
});
export type RegistryAuth = z.infer<typeof registryAuthSchema>;

export const registryHandshakeSchema = z.object({
  schemaVersion: z.string().min(1),
  protocolVersion: z.string().min(1),
  registryId: z.string().min(1),
  registryTitle: z.string().min(1),
  apiBaseUrl: z.string().min(1),
  auth: registryAuthSchema,
  cliCompatibility: versionSelectorSchema,
  bundleFormat: z.literal("rendo-bundle.v1"),
  digestAlgorithm: z.literal("sha256"),
});
export type RegistryHandshake = z.infer<typeof registryHandshakeSchema>;

export const remoteSearchEntrySchema = z.object({
  kind: z.string().min(1),
  id: z.string().min(1),
  title: z.string().min(1),
  version: z.string().min(1),
  category: z.string().min(1),
  templateKind: templateKindSchema.optional(),
  templateRole: templateRoleSchema.optional(),
  official: z.boolean(),
});
export type RemoteSearchEntry = z.infer<typeof remoteSearchEntrySchema>;

export const remoteSearchResponseSchema = z.object({
  registry: z.object({
    id: z.string().min(1),
    protocolVersion: z.string().min(1),
  }),
  results: z.array(remoteSearchEntrySchema),
});
export type RemoteSearchResponse = z.infer<typeof remoteSearchResponseSchema>;

export const remoteInspectResponseSchema = z.object({
  registry: z.object({
    id: z.string().min(1),
    protocolVersion: z.string().min(1),
  }),
  payload: inspectPayloadSchema,
  manifest: templateManifestSchema.optional(),
  bundle: z
    .object({
      url: z.string().min(1),
      digest: digestSchema,
      templateDigest: digestSchema,
      bundleFormat: z.literal("rendo-bundle.v1"),
    })
    .optional(),
});
export type RemoteInspectResponse = z.infer<typeof remoteInspectResponseSchema>;
